### ISSUE TITLE
Prototypes for methods with default arguments are not generated [imported]

### ISSUE THREAD

**Issue by cmaglie** [arduino/Arduino#386]
This is [Issue 386](http://code.google.com/p/arduino/issues/detail?id=386) moved from a Google Code project.
Added by 2010-10-23T13:23:50.000Z by [andre...@gmail.com](http://code.google.com/u/101138032102755792048/).
Please review that bug for more context and additional comments, but update this bug.

Original labels: Type-Defect, Priority-Medium, Component-PreProcessor

### Original description

Steps to reproduce the issue:

STEP 1. Insert the following code into a blank sketch:

```C++
void setup() {
  test(42);
  test2(42);
}

void loop() {}
void test(int arg) {};
void test2(int arg = 0) {};
```

STEP 2: Click the verify button and observe the error message:

```
DefaultArgBugDemo.cpp: In function 'void setup()':
DefaultArgBugDemo:2: error: 'test2' was not declared in this scope
```

STEP 3:
Examine the generated *.cpp file and observe that a prototype has been generated for test1 but not test2:

```C++
#include "WProgram.h"
void setup();
void loop();
void test(int arg);
void setup() {
  test(42);
  test2(42);
}

void loop() {}
void test(int arg) {};
void test2(int arg = 0) {};
```

**What is the expected output? What do you see instead?**

You would expect a prototype to be generated for test2.

**What version of the Arduino software are you using? On what operating**
**system?  Which Arduino board are you using?**

Tested in 0018 and 0021. Running on Win7 Home Premium (though the issue likely exists on all platforms)

**Please provide any additional information below.**

Of course the workaround is relatively simple (just define the prototype yourself) but it would be nice for all methods to be treated equally.

**Comment by ffissore** [comment]
Please checkout this thread [1] on devs mailing list and try one of the linked IDEs. It should fix the problems reported in this issue
[1] https://groups.google.com/a/arduino.cc/forum/#!topic/developers/4X2T3rCxXWM

**Comment by ffissore** [comment]
With #3779 the preprocessed code is now 
```
#include <Arduino.h>
#line 1
#line 1 "sketch_sep11c.ino"
void setup();
void loop();
void test(int arg);
void test2(int arg = 0);
#line 1
void setup() {
  test(42);
  test2(42);
}

void loop() {}
void test(int arg) {};
void test2(int arg = 0) {};
```

which now fails with 
```
sketch_sep11c.ino: In function 'void test2(int)':
sketch_sep11c.ino:8:23: error: default argument given for parameter 1 of 'void test2(int)' [-fpermissive]
sketch_sep11c.ino:4:6: error: after previous specification in 'void test2(int)' [-fpermissive]
default argument given for parameter 1 of 'void test2(int)' [-fpermissive]
```

Can you help spot the error? Code looks good to me

**Comment by cmaglie** [comment]
The default argument must be defined only once per compilation unit, for example you can't write:

```C++
void test(int a=0);
void test(int a=0) { /* empty */ }
```

even if the default argument `= 0` is the same.
To fix it, it seems that you can arbitrarily choose to drop the default argument in the prototype or in the implementation, I've tried to compile with g++:

```C++
void test(int a);
void test(int a=0) { /* empty */ }
```

and

```C++
void test(int a=0);
void test(int a) { /* empty */ }
```

and both compiles fine.

**Comment by ffissore** [comment]
Now I remember an old chat about it. I chose not to handle default args. Reason is that the correct way of using default args is to have the prototype list them, as in @cmaglie second example

```C++
void test(int a=0);
void test(int a) { /* empty */ }
```

However this implies that preprocessor must **change** original source code: I mean not just inserting code, but changing existing code. This is dangerous as we don't have a source code manipulation tool powerful enough to handle all the possible cases (eg: functions defined on multiple lines, spaces, comments between arguments on multi lines...)

So this is a wontfix. If you want to use default arguments, you have to define the prototype yourself. #3779 is now smart enough to understand that a function with default args already has its prototype and leave it alone
